src/toc.tsx

```diff
@@ -110,13 +143,27 @@ export class TableOfContents extends Widget {
     });
   }
 
+  get generator() {
+    if (this._current) {
+      return this._current.generator;
+    }
+    return null;
+  }
+
   /**
    * Rerender after showing.
    */
   protected onAfterShow(msg: Message): void {
     this.update();
   }
 
+  set notebookMetadata(value: [string, any]) {
```

Comment:
I'd like to here more about what the purpose of setting the notebook metadata is. Is the metadata meant to be useful outside of the TOC context? Is it something that we could accomplish in another way?

Reply:
These metadata just records what toolbar button user toggled (for example, if they toggle code cells off, next time they will not see the code cells in the document by default). I used to think of storing these data in LayoutRestorer or the user settings of the JupyterLab, but the reason that made me stored them in the notebook metadata in the end is these settings varies in different notebooks. Users might have code cells toggled on in one notebook but toggled off in another notebook.

Reply:
Interesting. That makes some sense to me. Maybe we could remove the `NotebookTracker` from the TOC by giving the tracker to the toolbar which is generated. Then the notebook metadata could be set via the various button presses, rather than routing through the TOC widget.

Reply:
Yes this is what I already changed!

Reply:
Great, thanks!

Reply:
I have you pushed that commit? I don't see it here...

Reply:
I pushed the commit which caused the TOC not show up in notebook (which is the first issue), but I'm sorry I didn't push the commits for metadata refactoring because I'm working on storing the toolbar instead of re-creating it every time. I'm pushing the metadata refactoring right now.